[Woo POS][Settings] i2: Card reader management design updates

iamgabrielma opened the pull request:
Closes WOOMOB-1646

## Description
This PR updates the card reader "management" section of POS settings by updating the design following qKAWGmvUsvfnW0Z4CssJHe-fi and adding actions for disconnecting the card reader, as well as updating its software.

## Changes
- Updates card reader "management" view to latest designs
- Adds disconnect action from settings
- Adds update firmware action from settings
- Conditionally renders banner when firmware update is needed
- Updates card reader service, facade and reader model to account for firmware status and version
- Update formattedBatteryLevel to account for localization/LTR

## Test Steps
- Navigate to POS Settings
- Play with connecting the reader, disconnecting it, and attempt software updates by updating `Terminal.shared.simulatorConfiguration.availableReaderUpdate`.

## Screenshots

https://github.com/user-attachments/assets/18fc17b5-24c4-4ed0-bda2-051e6d110d6a

iamgabrielma committed:
extract POSInformationCard and POSInformationCardFieldRow

iamgabrielma committed:
add optional button to POSInformationCardFieldRow

iamgabrielma committed:
disconnect reader action

iamgabrielma committed:
update card reader facade and service with update method

iamgabrielma committed:
pass software version from cardreader to service

iamgabrielma committed:
pass software update availability state to POS

iamgabrielma committed:
add button style to POSInformationCardFieldRow

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
add close button to readerUpdateCompletion

In POSSettingsHardwareDetailView we disable dismissal of PointOfSaleCardPresentPaymentAlert via  `posInteractiveDismissDisabled`, which causes the merchant not being able to close the modal once a software update has finished. In order to keep this dimissal enabled for each case, we add an x button to close/dismiss the modal into the specific   PointOfSaleCardPresentPaymentReaderUpdateCompletionView view

iamgabrielma committed:
render conditional card when software update is needed

iamgabrielma committed:
make simpler POSSettingsCardWithIcon

iamgabrielma committed:
localization

iamgabrielma committed:
lint

iamgabrielma committed:
make test target compile

iamgabrielma committed:
update formattedBatteryLevel

iamgabrielma commented on Modules/Sources/PointOfSale/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -2,6 +2,7 @@ import CocoaLumberjackSwift
 import Foundation
 import Combine
 import Observation
+import enum Hardware.CardReaderSoftwareUpdateState
```
Thoughts on importing `Hardware` here? Should we keep it in Yosemite and inject it through somewhere else into the aggregate model? 🤔

joshheald replied:
You can import it from Yosemite as there's a typealias already in Model.swift
We shouldn't import Hardware above Yosemite
```suggestion
import enum Yosemite.CardReaderSoftwareUpdateState
```

iamgabrielma committed:
fix test

joshheald commented:
The buttons don't match designs, AFAICT:

Here's the button in designs: qKAWGmvUsvfnW0Z4CssJHe-fi-6633_5749

![IMG_ED948AF728AD-1](https://github.com/user-attachments/assets/dace6ef6-8760-4f07-94b1-b8e785ee91b9)

- Internal padding looks off
- Font should be smaller and bold
- Buttons should be vertically centred in their rows

joshheald commented on Modules/Sources/PointOfSale/Card Present Payments/CardPresentPaymentFacade.swift:
```diff
@@ -1,6 +1,7 @@
 import Foundation
 import enum Yosemite.PaymentChannel
 import struct Yosemite.Order
+import enum Hardware.CardReaderSoftwareUpdateState
```
This one can come from yosemite as well

joshheald commented on Modules/Sources/PointOfSale/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -28,12 +29,20 @@ protocol PointOfSaleAggregateModelProtocol {
     private(set) var orderStage: PointOfSaleOrderStage = .building
 
     private(set) var cardReaderConnectionStatus: CardPresentPaymentReaderConnectionStatus = .disconnected
+    private(set) var cardReaderUpdateState: CardReaderSoftwareUpdateState = .none
```
Not sure if it's a good idea, but potentially we could base this on cardPresentPaymentAlertViewModel instead of making a new subscription. But don't bother in this PR, maybe as a follow up and only if it's clean. There's nothing wrong with this approach.

joshheald commented on Modules/Sources/PointOfSale/Presentation/Settings/POSInformationCard.swift:
```diff
@@ -0,0 +1,92 @@
+                        .overlay {
+                            RoundedRectangle(cornerRadius: POSCornerRadiusStyle.small.value)
+                                .stroke(buttonStyle == .default ? Color.posOnSurface : Color.posPrimaryContainer, lineWidth: 2)
+                        }
```
This feels quite cluttered for a button. I think we should be using a defined style for the button , and making this a lot shorter in the process...

iamgabrielma replied:
Good point, I've extracted into a new button style on https://github.com/woocommerce/woocommerce-ios/pull/16321/commits/75aeb205930c865e50cb1fc356b9cdda28d3524d

joshheald commented on Modules/Sources/PointOfSale/Presentation/Settings/POSSettingsHardwareDetailView.swift:
```diff
@@ -146,25 +157,33 @@ private extension POSSettingsHardwareDetailView {
+                                                    subtitle: Localization.updateFirmwareBannerSubtitle)
+                        }
+                        POSInformationCard {
+                            VStack(spacing: POSSpacing.medium) {
```
Spacing looks off in this view (1), and so does the bottom padding (2):
![IMG_ED948AF728AD-1](https://github.com/user-attachments/assets/62600da0-0946-43db-a5e7-dd0e18051dab)

I think it might be due to too much padding in the cards?

iamgabrielma replied:
Yes, I was adding additional external padding when is already handled internally. Fixed!

<img width="2266" height="1488" alt="Simulator Screenshot - iPad mini (A17 Pro) - US store - 2025-11-07 at 09 43 37" src="https://github.com/user-attachments/assets/b23aed20-13fb-488e-816b-bfb29162a3f3" />

joshheald commented on Modules/Sources/PointOfSale/Presentation/Settings/POSSettingsHardwareDetailView.swift:
```diff
@@ -146,25 +157,33 @@ private extension POSSettingsHardwareDetailView {
+                                                           buttonStyle: .primary)
                             }
-                            .padding()
+                            .padding(.bottom, POSPadding.medium)
```
Or maybe it's this padding. Not sure, but needs a look...

joshheald commented on Modules/Tests/PointOfSaleTests/Mocks/MockCardPresentPaymentService.swift:
```diff
@@ -2,6 +2,7 @@ import Foundation
 import Combine
 import enum Yosemite.PaymentChannel
 import struct Yosemite.Order
+import enum Hardware.CardReaderSoftwareUpdateState
```
```suggestion
import enum Yosemite.CardReaderSoftwareUpdateState
```

joshheald commented on WooCommerce/Classes/POS/Adaptors/Card Present Payments/CardPresentPaymentService.swift:
```diff
@@ -6,13 +6,16 @@ import struct Yosemite.CardPresentPaymentsConfiguration
 import struct Yosemite.CardReader
 import enum Yosemite.CardPresentPaymentAction
 import enum Yosemite.PaymentChannel
+import enum Hardware.CardReaderSoftwareUpdateState
```
```suggestion
import enum Yosemite.CardReaderSoftwareUpdateState
```

joshheald commented on WooCommerce/Classes/POS/Adaptors/Card Present Payments/CardPresentPaymentService.swift:
```diff
@@ -211,6 +227,19 @@ private extension CardPresentPaymentService {
         }
     }
 
+    @MainActor
+    static func createUpdateStatePublisher(stores: StoresManager) async -> AnyPublisher<CardReaderSoftwareUpdateState, Never> {
+        return await withCheckedContinuation { continuation in
+            var nillableContinuation: CheckedContinuation<AnyPublisher<CardReaderSoftwareUpdateState, Never>, Never>? = continuation
+
+            let action = CardPresentPaymentAction.observeCardReaderUpdateState { updateStatePublisher in
+                nillableContinuation?.resume(returning: updateStatePublisher)
+                nillableContinuation = nil
+            }
+            stores.dispatch(action)
+        }
+    }
+
```
Potentially reducing these checked continuations and publishers by sharing/filtering the existing ones could help for the future. Again, this works, but it does seem like we're doing something again which we've already done for the main flow...

joshheald reviewed (approved):
Various changes to match designs required... I also wonder if we can avoid making new publishers/observables for this, or at least filtering the existing ones as a new observable...

iamgabrielma committed:
Merge branch 'trunk' into task/WOOMOB-1646-pos-settings-card-reader-management

iamgabrielma committed:
update imports layer for CardReaderSoftwareUpdateState

iamgabrielma committed:
fix too much horizontal padding

iamgabrielma committed:
update divider vertical padding

iamgabrielma committed:
adjust button style

iamgabrielma committed:
fix button vertical miss-alignment

iamgabrielma committed:
extract to new POS button style

iamgabrielma committed:
remove unused flag for pointOfSaleSettingsCardReaderFlow

iamgabrielma committed:
periphery: remove unused

we moved to use the button size on 75aeb205930c865e50cb1fc356b9cdda28d3524d rather than a fixed radius for all of them

iamgabrielma committed:
fix failing test

We created a new publisher for .observeCardReaderUpdateState, when dispatched it waits for the completion handler to be called before resuming, but the test wasn’t handling this new one.

iamgabrielma commented:
Thanks for the review, and the pointer to `Yosemite.CardReaderSoftwareUpdateState` 🙇 

Along with the design changes I've also removed an unused feature flag on https://github.com/woocommerce/woocommerce-ios/pull/16321/commits/8d3389d46849f062ca8e486f8f328bc1a1db5b69 and fixed a failing test on https://github.com/woocommerce/woocommerce-ios/pull/16321/commits/a3356208fc92ac57e4caf9e3bb7b6d8771a231d4

I'll look into improving the publisher separately on WOOMOB-1654 👍
